### PR TITLE
Issue #3700: Control Characters are not skipped with google_checks co…

### PR DESCRIPTION
**Pull request by Luolc**
…nfig

#3700 

Changes are according to https://en.wiktionary.org/wiki/Appendix:Control_characters.

Apart from changing 1 to 9 to cover [C0 (ASCII and derivatives)](https://en.wiktionary.org/wiki/Appendix:Control_characters#C0_.28ASCII_and_derivatives.29), I add a missing `00AD` from first row of [Unicode control characters](https://en.wiktionary.org/wiki/Appendix:Control_characters#Unicode_control_characters).

Diff report
http://www.luolc.com/checkstyle-diff-report/issue3700/

### CODE REVIEW

**Comment by Vladlis** [comment]
1) Please update com.google.checkstyle.test.chapter2filebasic.rule233nonascii.AvoidEscapedUnicodeCharactersTest

2) Please provide diff report and analyze it
https://github.com/checkstyle/contribution/tree/master/checkstyle-tester

**Comment by Luolc** [comment]
@Vladlis I have no experience of writing test for checkstyle and make diff report before. It might take me some time to view the guide document at first. Thanks for pointing it out!

**Comment by Luolc** [comment]
@rnveach Understood. Is there any specific rules to follow or recommended sample to write UT for checkstyle?

**Comment by rnveach** [comment]
> what is it folder for? Some special kind (google style?) of tests?

It is more of an integration folder. We use it showcase our google configuration to users who want to use it. It shows how we are obeying the google style guide and that users can use our utility and configuration in their projects and be compliant with google.

> Is there any specific rules to follow or recommended sample to write UT for checkstyle?

99% of the time we write tests like https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheckTest.java#L166-L188 .
We usually only break from this if there is something easier, like accessing a public method directly, or using reflection to reach the impossible areas.

**Comment by Luolc** [comment]
@rnveach Got it. 
smth strange is that the source code for the test is a little bit weird. like https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java#L58, there are tabs in the code and the indentation is in a mess, to some extent.

How should I write the source code in this situation? I think I need to list various of characters to verify them all. Is it OK to just simply write them down in a list or array? or smth else, whatever. That's what I really want to ask, the rule of writing sources.

**Comment by Luolc** [comment]
diff report here. http://www.luolc.com/checkstyle-diff-report/
it takes much time cloning projects, I will test more of them later.

I will go cycling at weekend and back at Monday :)

**Comment by rnveach** [comment]
> How should I write the source code in this situation? 

We do not really care about the formatting of inputs. We don't verify them for style standards.
Our only request is they clearly and minimally show the problem and/or possible solution.
In this situation we want a list of characters with either `ok` or `violation` comment after them.

> Is it OK to just simply write them down in a list or array?

There should be only 1 character per line so it is clear which one is the violation or not.
Anything else I will leave up to you for now. I just expect to see an easy to read list that is clear and concise.

> it takes much time cloning projects

This is only a 1 time process. Every next run won't clone the projects unless you delete them.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java [hunk 1]
```diff
@@ -102,4 +102,17 @@ void foo() {
 	private String notAUnicodeEscaped2 = "\\\\u1234";
 
 	private String onlyEscaped = "\\\u1234";
+
+	private String controlC0 = ""
+			+ "\u0000\u0001\u0002\u0003\u0004\u0005\u0006\u0007\u0008\u0009\n\u000b\u000c\r\u000e\u000f"
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: We need clear identification of which ones are good and bad. Violation just shows the line number and not the column number.
It is not clear if all are bad or just one is bad and the others are accidentally good. It won't be clear if we change expression in the future and mess up one of the characters.
Input should also have a comment identifying which are good or bad as a secondary verification.

@romani Please confirm the range of numbers is ok.
Should we expand this to every number between 0000 and FFFF?

File: src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java [hunk 2]
```diff
@@ -17,15 +17,15 @@ public Object fooString()
         String fakeUnicode = "asd\tsasd"; 
         String fakeUnicode2 = "\\u23\\u123i\\u";
         String content = "";
-        return "\ufeff" + content ; // byte order mark ok
+        /*byte order mark ok*/return "\ufeff" + content ;
```
Comment (rnveach) [inline src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java]: Is there a reason for this change?
We don't expect any unnecessary changes.

File: src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java [hunk 3]
```diff
@@ -17,15 +17,15 @@ public Object fooString()
...
     }
 
     public Object fooChar()
     {
         /*warn*/char unitAbbrev2 = '\u03bc'; 
         char unitAbbrev3 = '\u03bc'; // Greek letter mu, "s" ok
         String content = "";
-        return '\ufeff' + content; // byte order mark ok
+        /*byte order mark ok*/return '\ufeff' + content;
```
Comment (rnveach) [inline src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java]: same as above.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java (hunk 1 above)
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: Is there a reason for `\n` and `\r` and not their unicode equivalent like the others?

Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: I completely support rnveach comments.

> Should we expand this to every number between 0000 and FFFF or is that too much?

this issue is already a sign that we failed to control logic if tests are defined in ranges and not all are covered.
lets generate input file with all characters and clearly define violation for each unicode.

File: src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java (hunk 2 above)
Comment (Luolc) [inline src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java]: According to Google Style, characters with a tail comment and non-printable ones are both valid. We cannot determine which configuration plays a role using the original codes. Assuming that Google no longer allows the use of non-printable characters one day in the future (I know the possibility is very low), the test will still pass for the tail comments. In that case, we are missing the validation of non-printable characters in integration test. In other word, if our check of non-printable characters raise a false negative, we are not able to find it via this test.

File: src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java (hunk 3 above)
Comment (Luolc) [inline src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java]: Answered at above.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java (hunk 1 above)
Comment (Luolc) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: The missing unicode above are `\u000a` (`\n`) and `\u000d` (`\r`). In Java, unicode escapes are pre-processed before the compiler is run. Therefore, if you put `\u000a` in a String literal like 
```
String someString = "foo\u000Abar";
```
It will be compiled exactly as if you wrote
```
String someString = "foo
bar";
```
This can not even be compiled successfully. Unlike other common control characters (i.e. a Greek letter mu), these two may lead to more problems like compilation error. [Google Style](http://checkstyle.sourceforge.net/reports/google-java-style-20160712.html#s2.3.3-non-ascii-characters) said `The choice depends only on which makes the code easier to read and understand`. IMO, we shouldn't allow user to use these two characters. Google Style allows using escaped characters of non-printable ones, but not misleading ones. And in this case the escaped format of line feed and carriage return can be `\n` and `\r`, instead of `\u000a` and `\u000d`.

Comment (Luolc) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: Got it. That make sense.
I will write a new input file listing all control characters. Is it OK to write a new test method in https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheckTest.java to test all control characters specifically?

Comment (Luolc) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: I made an update but not not sure if it is a good practice. Need suggestions. Thanks!

File: src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java (hunk 2 above)
Comment (rnveach) [inline src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java]: Just to confirm, this change is because if `feff` is marked as bad, the test won't trigger a violation on this line because it has a comment afterwards, but with a comment before, it will?
If so, then I think it is fine.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java (hunk 1 above)
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: @Luolc Is there anyway to use the unicode equivalent without a compiler error?
The point of this check and this test is verify unicode ranges. We should have all in the range if possible.
If it is not possible at all in the source code, then we should probably add a comment so it is clear why they are not in the list in their specific position.

File: src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java (hunk 2 above)
Comment (Luolc) [inline src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java]: Yep. Thats why I make this change. If `feff` is marked as bad, a violation can be triggered like the line here https://github.com/checkstyle/checkstyle/blob/master/src/it/resources/com/google/checkstyle/test/chapter2filebasic/rule233nonascii/InputAvoidEscapedUnicodeCharacters.java#L15

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java (hunk 1 above)
Comment (Luolc) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: In some extreme cases.
For example,
```
String[] strs = new String[] {"\u000a"};
```
will be compiled as if you wrote
```
String[] strs = new String[] {
};
```
There will be no compiler error. But... obviously very misleading.
Should we take the cases like this into account?

Comment (Luolc) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: Emm... the example above is not correct. It should be
```
String[] strs = new String[] {"
"};
```
And there will still be a compiler error.
Let me think again. A sec

Comment (Luolc) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAvoidEscapedUnicodeCharacters.java]: It seems that it is impossible to use it in string literal without compiler error. We cannot split a string literal into two lines. It will surely produce an `illegal line end in character literal` error.
smth like
```
\\ there will be a line feed at the end of this comment. \u000a
```
may be compiled without error, while we only concern about string literal and char literal in this check (the accepted token types). To sum up, it is impossible.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAllUnicodeControlCharaters.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            + "\u0001"
+            + "\u0002"
+            + "\u0003"
+            + "\u0004"
+            + "\u0005"
+            + "\u0006"
+            + "\u0007"
+            + "\u0008"
+            + "\u0009"
+            + "\n"
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAllUnicodeControlCharaters.java]: Let's add a comment that says it is impossible to add the unicode for these.

Comment (Luolc) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAllUnicodeControlCharaters.java]: Done.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAllUnicodeControlCharaters.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * C0 control codes, in the Unicode range U+0000-U+001F.
+     * Inherited from the ASCII standard.
+     *
+     * Violation, when allowEscapesForControlCharacters is false
+     * OK, when allowEscapesForControlCharacters is true
+     */
+    private String controlC0 = ""
+            + "\u0000"
```
Comment (Luolc) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAllUnicodeControlCharaters.java]: All unicodes from 0x0000 to 0xffff are included in the input file.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAllUnicodeControlCharaters.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+ * @author LuoLiangchen
+ * @since 2017/3/7
+ */
+public class InputAllUnicodeControlCharaters {
+
+    /**
+     * C0 control codes, in the Unicode range U+0000-U+001F.
+     * Inherited from the ASCII standard.
+     *
+     * Violation, when allowEscapesForControlCharacters is false
```
Comment (Luolc) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAllUnicodeControlCharaters.java]: Every line is commented with `violation` or `ok` now.

**Comment by MEZk** [comment]
@rnveach @Luolc 
Personally, I would suggest rewriting the test to make it more clear and verbose:

```java
@Test
public void testAllowEscapesForControlCharacterSetForAllCharacters() throws Exception {
    final DefaultConfiguration checkConfig =
        createCheckConfig(AvoidEscapedUnicodeCharactersCheck.class);
    checkConfig.addAttribute("allowEscapesForControlCharacters", "true");

    final int indexOfStartLineInInputFile = 6;
    final String message = getCheckMessage(MSG_KEY);
    final String[] expectedViolations = IntStream.rangeClosed(0, 0xffff)
        .parallel()
        .filter(val -> !isControlCharacter(val))
        .mapToObj(msg -> indexOfStartLineInInputFile + msg + ": " + message)
        .toArray(String[]::new);

    verify(checkConfig, getPath("InputAllEscapedUnicodeCharacters.java"), expectedViolations);
}

private static boolean isControlCharacter(final int character) {
    return Arrays.binarySearch(C0_CONTROL_CHARACTER_INDICES, character) >= 0
        || Arrays.binarySearch(C1_CONTROL_CHARACTER_INDICES, character) >= 0
        || Arrays.binarySearch(OTHER_CONTROL_CHARACTER_INDICES, character) >= 0;
}
```
C0_CONTROL_CHARACTER_INDICES, C1_CONTROL_CHARACTER_INDICES, and OTHER_CONTROL_CHARACTER_INDICES may become arrays of chars (char[]). However, it will require to make type cast from int to char in the 
```
.filter(val -> !isControlCharacter((char) val))
```

and change the type of isControlCharacter parameter. So may not be required to do.
Feel free to agree or disagree. @Luolc I'm not forcing you to make changes. It is just a proposal. On my PC the performance is not severely damaged. The most time-consuming part is in verify method. At least [this loop](https://github.com/checkstyle/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/test/java/com/puppycrawl/tools/checkstyle/BaseCheckTestSupport.java#L260-L260) takes O(n).

**Comment by Luolc** [comment]
@MEZk updated. thanks a lot! That's much clearer than before. 
Actually I would prefer Java 8 style in my daily coding as well, but I am not sure if it is OK to use Java 8 grammar in checkstyle project before. I was worried about the compatibility. Therefore I wrote it in a for-loop way. 
Instead of the performance, the optimization of readability by using Java 8 style is much significant. I have printed the time cost when using for-loop. It takes about 30 ms in average while verify takes about 5000 ms :)

BTW, the CI keeps failing at the third one from last in Travis CI, report is like https://travis-ci.org/checkstyle/checkstyle/jobs/208912195. It is the only failure in that build. I am using mac as well, and running the same command on my local won't produce an error. I don't know why it will raise a stackoverflow error at https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java#L168 in that build job. Many other jobs are running the same test without failure. Need some help :)

**Comment by MEZk** [comment]
@Luolc 
> Actually I would prefer Java 8 style in my daily coding as well, but I am not sure if it is OK to use Java 8 grammar in checksty.le project before.

Java 8 is OK except for some imports. To become more familiar with the problems we faced take a look at 
1) https://github.com/checkstyle/checkstyle/issues/3843
2) https://github.com/checkstyle/checkstyle/pull/3853
3) https://github.com/checkstyle/checkstyle/pull/3849

**Comment by rnveach** [comment]
> BTW, the CI keeps failing at the third one from last in Travis CI, report is like 

I would just skip your file from that test.
We should probably create an issue trying to prevent that test from using the stack for very deep trees.

**Comment by Luolc** [comment]
@MEZk 
Thanks! I will have a look at the previous issues. 😄 
CI passed after I optimized the input file. Now there should be about 256 recursive call at most, as stackoverflow probably appears when calling over 1023 times in current CI's config according to the log here https://travis-ci.org/checkstyle/checkstyle/jobs/208912195#L254. In addition, the running time of AvoidEscapedUnicodeCharactersCheckTest also reduced about a half.

@rnveach 
The optimization of input file does work. CI passed and the running time also reduced.
>  It should be possible to rewrite the test to not use the stack, but then we may use just as much memory.

Agree. A non-recursive way can be better.

Feel free to point out anything else I can do for this PR. 😄

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheckTest.java [hunk 7]
```diff
@@ -32,6 +34,109 @@
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
 
 public class AvoidEscapedUnicodeCharactersCheckTest extends BaseCheckTestSupport {
+
+    private static final int[] C0_CONTROL_CHARACTER_INDICES = {
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheckTest.java]: please make comment above each such  array to define from where values are taken.
Like: // C0 (ASCII and derivatives) - https://en.wiktionary.org/wiki/Appendix:Control_characters#C0_.28ASCII_and_derivatives.29

please do for that for all other arrays.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java [hunk 8]
```diff
@@ -120,8 +120,8 @@
 
     /** Regular expression Unicode control characters. */
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java]: please add link to wiki page here too, to define source of control codes.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheckTest.java (hunk 7 above)
Comment (Luolc) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheckTest.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java (hunk 8 above)
Comment (Luolc) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java]: Done.
